test/e2e/cht-docker-utils.js

```diff
@@ -0,0 +1,135 @@
+      file.on('error', () => file.close(reject));
+    })
+    .on('error', () => {
+      fs.unlinkSync(file.path);
+      file.close(() => reject('Failed to download CHT Docker Helper script "cht-docker-compose.sh"'));
+    });
+});
+
+const ensureScriptExists = async () => {
+  if (!fs.existsSync(dockerHelperDirectory)) {
+    await fs.promises.mkdir(dockerHelperDirectory);
+  }
+
+  if (!fs.existsSync(dockerHelperScript)) {
+    await downloadDockerHelperScript();
+  }
+};
+
+const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));
+
+const getProjectConfig = async (projectName) => {
+  try {
+    const configFile = await fs.promises.readFile(path.resolve(dockerHelperDirectory, `${projectName}.env`), 'utf8');
+    return Object.fromEntries(
+      configFile.toString()
+        .split('\n')
+        .map(line => line.split('='))
+        .filter(entry => entry.length === 2),
+    );
+  } catch (error) {
+    log.error(error);
+    return {
+      COUCHDB_USER: 'medic',
+      COUCHDB_PASSWORD: 'password',
+      NGINX_HTTPS_PORT: '10443',
+    };
+  }
+};
+
+const getProjectUrl = async (projectName = DEFAULT_PROJECT_NAME) => {
+  const config = await getProjectConfig(projectName);
+  const { COUCHDB_USER, COUCHDB_PASSWORD, NGINX_HTTPS_PORT } = config;
+  return `https://${COUCHDB_USER}:${COUCHDB_PASSWORD}@127-0-0-1.local-ip.medicmobile.org:${NGINX_HTTPS_PORT}`;
+};
+
+const isProjectReady = async (projectName, attempt = 1) => {
+  log.info(`Checking if CHT is ready, attempt ${attempt}.`);
+  const url = await getProjectUrl(projectName);
+  await request({ uri: `${url}/api/v2/monitoring`, json: true })
+    .catch(async (error) => {
+      if (
+        error.error.code !== 'DEPTH_ZERO_SELF_SIGNED_CERT' ||
+        ![502, 503].includes(error.statusCode)
+      ) {
+        // unexpected error, log it to keep a trace,
+        // but we'll keep retrying until the instance is up, or we hit the timeout limit
+        log.trace(error);
+      }
+
+      await sleep(1000);
+      return isProjectReady(projectName, attempt + 1);
+    });
+};
+
+const startProject = (projectName) => new Promise((resolve, reject) => {
+  const configFile = path.resolve(dockerHelperDirectory, `${projectName}.env`);
+  if (fs.existsSync(configFile)) {
+    // project config already exists, reuse it
+    const childProcess = spawn(dockerHelperScript, [`${projectName}.env`, 'up'], { cwd: dockerHelperDirectory });
+    childProcess.on('error', reject);
+    childProcess.on('close', resolve);
+  } else {
+    // initialize a new project, config will be saved to `${projectName}.env`
+    // stdio: 'pipe' to answer the prompts to initialize a project by writing to stdin
+    const childProcess = spawn(dockerHelperScript, { stdio: 'pipe', cwd: dockerHelperDirectory });
+    childProcess.on('error', reject);
+    childProcess.on('close', async () => {
+      await isProjectReady(projectName);
+      resolve();
+    });
+
+    childProcess.stdin.write('y\n');
```

Comment:
Should we maybe log a cht-core ticket to enhance the docker-helper to support taking arguments on the initial command instead of requiring interactive prompts?  This  works fine here for now, but will fail if prompts are added/removed from the docker-helper.  (Also, being able to pass everything as args would be generally useful for the docker-helper since it would make it easier to script with...)

Reply:
I was thinking about this to! While it'd be nice if the helper functionality was in `cht-conf` as has been kicked around, instead:

I wondered about a docker helper refactor that would break out all the functions, but specifically the `init_env_file()` [function](https://github.com/medic/cht-core/blob/e6d184946affc62773d569168216a5b913f38a30/scripts/docker-helper-4.x/cht-docker-compose.sh#L17) and `create_compose_files()`  [function](https://github.com/medic/cht-core/blob/e6d184946affc62773d569168216a5b913f38a30/scripts/docker-helper-4.x/cht-docker-compose.sh#L66C1-L66C21) into a stand alone "docker helper function" file.  These could then be included in docker helper with effectively no code changes while allowing external scripts (like this e2e test) use a loose API of sorts to legally, but programmatically call the to exact functions.  As it's a loose API, we can ensure a bit of backwards comparability going forwards.

If this sounds helpful and viable - happy to jump on this refactor!

Reply:
Honestly I love the idea! To push it further, while working on this setup, I was contemplating how convenient it would be to have a javascript _script_ instead of a bash script. Instead of having a bunch of `child_process.spawn('cht-docker-compose.sh')` to interface with the script, we could call javascript functions to orchestrate all this.

Technically it wouldn't change anything for app devs, instead of running `./cht-docker-compose.sh` they would run `./cht-docker-compose.js`. As long as the file has the execute permission `+x` and a shebang at the top `#!/usr/bin/env -S node`, it will work fine

For this PR specifically I don't think we should make any change but we can already open a cht-core issue to track it and discuss it there

Reply:
https://github.com/medic/cht-core/issues/9261